### ISSUE TITLE
[BUG] wrong signature for `PyFloat_FromString`

### ISSUE THREAD

**Issue by dvarrazzo** [cython/cython#3909]
The `PyFloat_FromString` function has the wrong signature.

https://github.com/cython/cython/blob/ed25f12b9e1fd9b7d5c3af56a9e61fd68d1a1d11/Cython/Includes/cpython/float.pxd#L24

This signature was for Python 2. [In Python 3](https://docs.python.org/3.5/c-api/float.html#c.PyFloat_FromString) it is:
```
 PyObject* PyFloat_FromString(PyObject *str)
```

the `pend` argument has gone.

**Comment by da-woods** [comment]
For the moment we still support Python 2 so I don't think we can just change it. It's simplified slightly since the second option is ignored anyway. Two options are:

1. Use the varargs syntax
    ```
    object PyFloat_FromString(object str, ...)
    ```
    This makes it the users' problem to pick the correct number of extra arguments for the Python version they're using (but the code they write will be version specific)

2. Have Cython define a (one argument) wrapper macro and divert all calls though that.

Option 1 has the advantage of not breaking any existing code.

**Comment by dvarrazzo** [comment]
Couldn't it be ifdeffed?

https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html?highlight=preprocessor#conditional-statements

**Comment by da-woods** [comment]
> Couldn't it be ifdeffed?

I don't think Cython defines a constant for the Python version. The more general problem with that (and probably all the compile-time conditional statements) is that we don't assume that the Python version used to run Cython is the same that the module will be compiled with.

**Comment by dvarrazzo** [comment]
Do you have any suggestion about how to write code that needs to support Python 3 only? Python 2 was deprecated one year ago.

**Comment by da-woods** [comment]
Yes - the `cpython cimport`s that Cython provides are for convenience only. You can can just duplicate the bits you need yourself. At the top of your pyx file:

```
cdef extern from "Python.h":
    object PyFloat_FromString(object)
```

is literally all you need.

-----------------

Updating the version that's provided with Cython is a little trickier because Cython hasn't yet dropped support for Python 2.

**Comment by dvarrazzo** [comment]
[Looks good](https://github.com/psycopg/psycopg3/commit/8432e27db23f321857e44cb585078156a265c8ec), thank you!

**Comment by scoder** [comment]
Have you tried just calling `float()`? I doubt that it makes a visible difference.

**Comment by dvarrazzo** [comment]
Yes, if you see the diff linked above, `float()` is what I was using. I use cython to generate a C extension for Python the way I would have written manually but without the chore of dealing with the refcount myself. `float()` creates an intermediate `bytes` objects, with its memory allocation and memcopy. I'd rather not have it in the innermost loop of psycopg3.

**Comment by scoder** [comment]
> `float()` creates an intermediate bytes objects, with its memory allocation and memcopy.

Huh? But that's exactly what your code now does – manually. (`cdef bytes b = data`, where `data` is a `char*`). Calling `float()` would essentially do the same thing, just automatically.

If you want to avoid creating the intermediate `bytes` creation, I suggest using the more low-level function [`PyOS_string_to_double()`](https://docs.python.org/3/c-api/conversion.html#c.PyOS_string_to_double), which seems perfect for your purpose. I [just added the declaration](https://github.com/cython/cython/commit/b64eafe74baa69faf0457c981b0f0ea31b892286) for the next 0.29.22 release.

**Comment by dvarrazzo** [comment]
> Huh?

@scoder Yes, you are right, sorry: it was `int()` which was causing an extra `bytes`. `float()` calls through the `PyObject_AsDouble` dispatcher though, which can be avoided as I know that the argument is bytes.

> I suggest using the more low-level function

Thank you very much! I missed it, as it's not documented in the Python float API.

**Comment by dvarrazzo** [comment]
> I just added the declaration for the next 0.29.22 release.

Shouldn't it be marked as `except? -1`?

WIthout:

```
In [11]: _psycopg3.FloatLoader(0).load(b"1e500")
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
OverflowError: value too large to convert to float: '1e500'

The above exception was the direct cause of the following exception:

SystemError                               Traceback (most recent call last)
<ipython-input-11-ac6d8af9d109> in <module>
----> 1 _psycopg3.FloatLoader(0).load(b"1e500")

SystemError: <cyfunction CLoader.load at 0x7f6a9430c790> returned a result with an error set
```

With:

```
In [3]: _psycopg3.FloatLoader(0).load(b"1e500")
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<ipython-input-3-ac6d8af9d109> in <module>
----> 1 _psycopg3.FloatLoader(0).load(b"1e500")

~/dev/psycopg3/psycopg3_c/psycopg3_c/adapt.pyx in psycopg3_c._psycopg3.CLoader.load()

~/dev/psycopg3/psycopg3_c/psycopg3_c/types/numeric.pyx in psycopg3_c._psycopg3.FloatLoader.cload()

OverflowError: value too large to convert to float: '1e500'
```

**Comment by scoder** [comment]
Right. Thanks for catching that. Fixed in 09ccf3c23b58d44f96852d971b43974a02f591fb.

**Comment by scoder** [comment]
Let's change the signature in 3.0. It's the right time to break backwards compatibility for a good cause.
